Book clubs update and search features

ChrisThackrey opened the pull request:
-- Finished Search Bar Functionality
-- Linked main bookclub view with details view
-- Added actions and reducer to AppContext
-- Implemented Fuzzy Search Enhancement
-- other small stuff

jfleming9357 commented on client/components/bookClubPage/BookClubItem.jsx:
```diff
@@ -1,29 +1,25 @@
-      <Container className='my-1 p-1'>
-        {/* <Card.ImgOverlay>
+    <Container className='my-1 p-1' onClick={handleClicked}>
+      {/* <Card.ImgOverlay>
```
Should the click event be on the container or the image? When I click anywhere on the carousel I get 404 errors for trying to get /bookclubs/undefined

ChrisThackrey replied:
change request handled

jfleming9357 commented on client/context/context.jsx:
```diff
@@ -1,11 +1,240 @@
+      dispatch({
+        type: 'LEAVE_BOOKCLUB',
+        payload: user._id
+      });
```
Oops wrong spot throwing the error. The above two functions both go to the catch block, and then throw an error in that catch block of trying to read 'data' of undefined

ChrisThackrey replied:
Which one was this? Not using dispatch/reducer now. This one may not using the global state setup on Daniel's end, I think he was on join/leave functionality.

jfleming9357 reviewed (changes requested):
Some bugs that need to be fixed. I don't love the fact that state is getting copied and reset to update one field of state every time in the context file, but I don't know enough about the method you are using to say whether there is a better way to do it. Would it be more efficient to have a lot of useState hooks? There would be quite a lot but at least you wouldn't have to copy the entire state object. Again don't know what the best practice for this is, just a shot in the dark.

ChrisThackrey commented:
> Some bugs that need to be fixed. I don't love the fact that state is getting copied and reset to update one field of state every time in the context file, but I don't know enough about the method you are using to say whether there is a better way to do it. Would it be more efficient to have a lot of useState hooks? There would be quite a lot but at least you wouldn't have to copy the entire state object. Again don't know what the best practice for this is, just a shot in the dark.

Refactored to a bunch of useState's, removed useReducer, still works and doesn't copy entire state object.

jfleming9357 reviewed (approved):
Looks good! Thank you!
